crates/web-plugins/didcomm-messaging/message-api/src/lib.rs

```diff
@@ -0,0 +1,75 @@
+use std::{collections::HashMap, fmt::Debug, marker::PhantomData, sync::Arc};
+
+type MessageHandler<S, M, R> = fn(Arc<S>, M) -> Result<R, PluginError>;
+
+#[derive(Debug, PartialEq)]
+pub enum PluginError {
+    InitError,
+}
+
+#[derive(Debug, Clone)]
+pub struct MessageRouter<S, M, R>
+where
+    S: Clone + Sync + Send + 'static,
+    M: Send + 'static,
+    R: Send + 'static,
+{
+    routes: HashMap<String, MessageHandler<S, M, R>>,
+    _marker: PhantomData<(S, M, R)>,
+}
+
+impl<S, M, R> MessageRouter<S, M, R>
+where
+    S: Clone + Sync + Send + 'static,
+    M: Send + 'static,
+    R: Send + 'static,
+{
+    pub fn new() -> Self {
+        Self {
+            routes: HashMap::new(),
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn route(mut self, msg_type: &str, f: MessageHandler<S, M, R>) -> Self {
+        self.routes.insert(msg_type.to_string(), f);
+        self
+    }
+
+    pub fn merge(&mut self, other: &Self) {
```

Comment:
Ideally, this function should consume `self` and incorporate the other message router, allowing it to be chained and collect the other router within the same instance instead of invoking it multiple times.